pkg/rpc/rpc.go

```diff
@@ -0,0 +1,166 @@
+package rpc
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"go.uber.org/zap"
+	"io"
+	"sync"
+)
+
+type JsonRpcRequest struct {
+	Method string          `json:"method"`
+	Params json.RawMessage `json:"params"`
+	Id     uint64          `json:"id"`
+}
+
+type JsonRpcResponse struct {
+	Result json.RawMessage `json:"result"`
+	Error  string          `json:"error"`
+	Id     uint64          `json:"id"`
+}
+
+type Error struct {
+	cause error
+}
+
+func (err *Error) Error() string {
+	return fmt.Sprintf("RPC error: %s", err.cause.Error())
+}
+
+func (err *Error) Unwrap() error {
+	return err.cause
+}
+
+type RPC struct {
+	writer    io.Closer // use encoder for writing instead
+	encoder   *json.Encoder
+	encoderMu sync.Mutex
+
+	decoder *json.Decoder
+	logger  *zap.SugaredLogger
+
+	pendingRequests map[uint64]chan JsonRpcResponse
+	lastRequestId   uint64
+	requestsMu      sync.Mutex
+
+	errChannel chan struct{} // never transports a value, only closed through setErr() to signal an occurred error
+	rpcErr     *Error        // only initialised via setErr(), if a rpc (non-skippable) error has occurred
+	errMu      sync.Mutex
+}
+
+// NewRPC creates and returns an RPC object and triggers a go routine to assign incoming responses to given requests.
+// The returned object cannot be used further if an error has occurred and must be recreated.
+func NewRPC(writer io.WriteCloser, reader io.Reader, logger *zap.SugaredLogger) *RPC {
```

Comment:
Starting a go routine is more of an implementation detail and not really relevant for the caller. Also, I'd move the documentation for the error handling to the functions that can actually return an error.